src/de.rs

```diff
@@ -582,8 +677,68 @@ mod tests {
         enum K {
             E(E),
         }
+        assert_eq!(from_str_auto::<K>("d1:y1:E1:t1:N1:ci1ee")?, K::E(E::N(1)));
+        Ok(())
+    }
+
+    #[test]
+    fn test_auto() -> Ret {
+        #[derive(Debug, PartialEq, Eq)]
+        struct TestAuto {
+            string: String,
+            bytes: Box<[u8]>,
+        }
+
+        struct TestAutoVisitor;
+        impl<'de> Visitor<'de> for TestAutoVisitor {
+            type Value = TestAuto;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                todo!()
+            }
 
-        assert_eq!(from_str::<K>("d1:y1:E1:t1:N1:ci1ee")?, K::E(E::N(1)));
+            fn visit_map<A>(self, mut map: A) -> std::result::Result<Self::Value, A::Error>
+            where
+                A: de::MapAccess<'de>,
+            {
+                let mut string = None;
+                let mut bytes: Option<&[u8]> = None;
+
+                match map.next_key()?.unwrap() {
+                    "string" => string = Some(map.next_value()?),
+                    "bytes" => bytes = Some(map.next_value()?),
+                    _ => panic!(),
+                };
+                match map.next_key()?.unwrap() {
+                    "string" => string = Some(map.next_value()?),
+                    "bytes" => bytes = Some(map.next_value()?),
+                    _ => panic!(),
+                };
+                assert!(map.next_key::<()>()?.is_none());
+                Ok(TestAuto {
+                    bytes: bytes.unwrap().to_vec().into_boxed_slice(),
+                    string: string.unwrap(),
+                })
+            }
+        }
+
+        impl<'de> de::Deserialize<'de> for TestAuto {
+            fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                deserializer.deserialize_any(TestAutoVisitor)
+            }
+        };
+        let input = b"d6:string4:12345:bytes4:\x01\x02\x03\x04e";
```

Comment:
BTW, `b"\x01\x02\x03\x04"` is a valid UTF-8.  Just put `\xFF` at any place to make it invalid UTF-8 and make the test stricter.

Reply:
Hm, I thought control characters are invalid UTF-8. Great catch!

Reply:
You are welcome!

UTF-8 is backward compatible with `0x00..=0xFF` ASCII range.

Reply:
... `0x00..=0x7F` ASCII range.